source/grid/tria.cc

```diff
@@ -1760,6 +1760,448 @@ namespace internal
         }
     }
 
+
+
+    /**
+     * An interface for algorithms that implement Triangulation-specific tasks
+     * related to creation, refinement, and coarsening.
+     *
+     * @note It is necessary to write out the functions for all possible
+     *   dim/spacedim combinations here to be able to make the functions
+     *   virtual.
```

Comment:
Can you explain why we need to write all the functions and not just use the `dim, spacedim` parameters from the class in the `create_triangulation`? At least in the implementations right now and https://github.com/peterrum/dealii/blob/c6dc588e7b49e291bf3ab0ac9017e02d051ca6bb/source/grid/tria.cc I do not see what would not work with that approach. But there are lots of convolutions in that file, so I could easily miss it.

Reply:
I am not completely sure if I understand your question correctly. My goal is to make the function virtual so that I can switch policies (for creating the coarse grid and/or the refinement strategy). To be able make the functions vitrual, the functions cannot be templated. One way to do that would be to write out all possible dim-spacedim combinations as I did. Alternatively, I could use the template arguments of the Policy:

https://github.com/dealii/dealii/blob/fbadd26068f8e95c6100a2f7b23ae538be7e409c/source/grid/tria.cc#L1773-L1774

i.e.: write:
```cpp
     void
     Policy<dim, spacedim>::create_triangulation(const std::vector<Point<spacedim>> &   v,
                           const std::vector<CellData<dim>> &cells,
                           const SubCellData &             subcelldata,
                           Triangulation<dim, spacedim> &           triangulation);
```

But to be able to do that we would need to merge the different template specializations (for 1D, 2D, 3D).

The problem in merging the `create_triangulation()` functions (as they are now) to a single huge function is that we get `TriaAccessors`, which might be invalid (e.g. `structdim=2,dim=2,spacedim=1`) and might not have certain needed functions like `set_used_flag`, `at_boundary`, and many more.... 

The other function have problems due to `RefinementCase`, which is templated, and some other functions that do not exist for certain dimensions.

Reply:
I was thinking of your second way with the `dim` and `spacedim` parameters in `create_triangulation`. My idea would be to create specializations of `PolicyWrapper` (for the quad/hex case) for `1,spacedim`, `2,spacedim` and `3,spacedim`. Since this class only implements the parts that are actually different between `dim=1,2,3` and not the other part of `Triangulation` (which prevent the use of partial specializations in that class), I would have thought that one can use such a setup. As to refinement case, I think one could find similar solutions, but that's why I'm asking: I don't know how much of the code you would like to move into these functions and where things start to collide. All I was pointing out was that using the `dim,spacedim` parameter on a class and then provide functions which belong to a different dim/spacedim felt unnatural. In the worst case, we could have free functions to specialize.

Reply:
> In the worst case, we could have free functions to specialize.

As far as I remember free functions will not work because the static methods access private fields of `Triangulation` (enables by the `friend` key word).

Reply:
@kronbichler I was able to significantly simplify the code. Let's merge this. I have 1-2 follow-up PRs in the pipeline.